Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
+
+    CartesianPower(InputIt first_, InputIt last_, size_t power_);
+    template<typename Container>
+    CartesianPower(const Container& container, size_t power_);
```

Comment:
I don't know. With such guideline, it could become over-complicated: members should have `m` prefix, constructor parameters should have trailing `_`, other methods' arguments should have "normal" naming, but if the're setting one of members, one should then find a new name or use this ugly `this->member` notation.

The best guideline I've ever seen was adding trailing `_` to **all** members, both `private` and `public`. It could be `m` prefix as well. This way argument names will be consistent and they will differ from member names enough to notice, which variable is a member.